components/ai_chat/core/browser/engine/engine_consumer_conversation_api_unittest.cc

```diff
@@ -0,0 +1,243 @@
+          }));
+  run_loop3.Run();
+  testing::Mock::VerifyAndClearExpectations(mock_api_client);
+}
+
+TEST_F(EngineConsumerConversationAPIUnitTest,
+       GenerateEvents_HistoryWithSelectedText) {
+  // Tests events building from history with selected text and new query without
+  // selected text but with page association.
+  ConversationHistory history = {
+      {mojom::CharacterType::HUMAN, mojom::ActionType::QUERY,
+       mojom::ConversationTurnVisibility::VISIBLE,
+       "Which show is this catchphrase from?", "I have spoken."},
+      {mojom::CharacterType::ASSISTANT, mojom::ActionType::RESPONSE,
+       mojom::ConversationTurnVisibility::VISIBLE, "The Mandalorian.",
+       std::nullopt}};
+  std::string expected_events = R"([
+    {"role": "user", "type": "pageText", "content": "This is my page. I have spoken."},
+    {"role": "user", "type": "pageExcerpt", "content": "I have spoken."},
+    {"role": "user", "type": "chatMessage", "content": "Which show is this catchphrase from?"},
+    {"role": "assistant", "type": "chatMessage", "content": "The Mandalorian."},
+    {"role": "user", "type": "chatMessage", "content": "Is it related to a broader series?"}
+  ])";
+  auto* mock_api_client = GetMockConversationAPIClient();
+  base::RunLoop run_loop;
+  EXPECT_CALL(*mock_api_client, PerformRequest(_, _, _))
+      .WillOnce([&](const std::vector<ConversationEvent>& conversation,
+                    EngineConsumer::GenerationDataCallback data_callback,
+                    EngineConsumer::GenerationCompletedCallback callback) {
+        // Some structured EXPECT calls to catch nicer errors first
+        EXPECT_EQ(conversation.size(), 5u);
+        EXPECT_EQ(conversation[0].role, mojom::CharacterType::HUMAN);
+        EXPECT_EQ(conversation[0].type, ConversationAPIClient::PageText);
+        EXPECT_EQ(conversation[1].role, mojom::CharacterType::HUMAN);
+        EXPECT_EQ(conversation[2].role, mojom::CharacterType::HUMAN);
+        EXPECT_EQ(conversation[3].role, mojom::CharacterType::ASSISTANT);
+        EXPECT_EQ(conversation[4].role, mojom::CharacterType::HUMAN);
+        // Match entire JSON
+        EXPECT_STREQ(mock_api_client->GetEventsJson(conversation).c_str(),
+                     FormatComparableEventsJson(expected_events).c_str());
+        std::move(callback).Run("");
+      });
+  engine_->GenerateAssistantResponse(
+      false, "This is my page. I have spoken.", std::nullopt, history,
+      "Is it related to a broader series?", base::DoNothing(),
+      base::BindLambdaForTesting(
+          [&run_loop](EngineConsumer::GenerationResult) { run_loop.Quit(); }));
+  run_loop.Run();
+  testing::Mock::VerifyAndClearExpectations(mock_api_client);
+}
+
+TEST_F(EngineConsumerConversationAPIUnitTest, GenerateEvents_Rewrite) {
+  std::string expected_events = R"([
+    {"role": "user", "type": "userText", "content": "Hello World"},
+    {"role": "user", "type": "requestRewrite", "content": "Use a funny tone"}
```

Comment:
@evq this is one area that I don't think is covered by the existing events - rewriting text in place. This is what I came up with, let me know if you want to change the event format